server/service/src/processors/transfer/shipment/mod.rs

```diff
@@ -0,0 +1,224 @@
+            let operation = match &log.row_action {
+                ChangelogAction::Upsert => get_upsert_operation(&ctx.connection, &log)
+                    .map_err(Error::GetUpsertOperationError)?,
+                ChangelogAction::Delete => get_delete_operation(&ctx.connection, &log)
+                    .map_err(Error::GetDeleteOperationError)?,
+            };
+
+            let record = ShipmentTransferProcessorRecord {
+                operation,
+                other_party_store_id: active_stores
+                    .get_store_id_for_name_id(name_id)
+                    .ok_or_else(|| Error::NameIsNotAnAciveStore(log.clone()))?,
+            };
+
+            // Try record against all of the processors
+            for processor in processors.iter() {
+                processor
+                    .try_process_record_common(&ctx.connection, &record)
+                    .map_err(Error::ProcessorError)?;
+            }
+
+            key_value_store_repo
+                .set_i64(
+                    KeyValueType::ShipmentTransferProcessorCursor,
+                    Some(log.cursor + 1),
+                )
+                .map_err(Error::DatabaseError)?;
+        }
+    }
+
+    Ok(())
+}
+
+#[derive(Error, Debug)]
+pub(crate) enum GetUpsertOperationError {
+    #[error("Shipment not found {0:?}")]
+    ShipmentNotFound(ChangelogRow),
+    #[error("Database error while fetching shipment with id {0} {1:?}")]
+    DatabaseError(String, RepositoryError),
+}
+
+fn get_upsert_operation(
+    connection: &StorageConnection,
+    changelog_row: &ChangelogRow,
+) -> Result<Operation, GetUpsertOperationError> {
+    use GetUpsertOperationError::*;
+    let repo = InvoiceRepository::new(connection);
+
+    let shipment = repo
+        .query_one(InvoiceFilter::new_match_id(&changelog_row.record_id))
+        .map_err(|e| DatabaseError(changelog_row.record_id.clone(), e))?
+        .ok_or_else(|| ShipmentNotFound(changelog_row.clone()))?;
+
+    let linked_shipment = match &shipment.invoice_row.linked_invoice_id {
+        Some(id) => repo
+            .query_one(InvoiceFilter::new_match_id(id))
+            .map_err(|e| DatabaseError(id.to_string(), e))?,
+        None => repo
+            .query_one(InvoiceFilter::new_match_linked_invoice_id(
+                &shipment.invoice_row.id,
+            ))
+            .map_err(|e| DatabaseError(shipment.invoice_row.id.clone(), e))?,
+    };
+
+    Ok(Operation::Upsert {
+        shipment,
+        linked_shipment,
+    })
+}
+
+fn get_delete_operation(
+    connection: &StorageConnection,
+    changelog_row: &ChangelogRow,
+) -> Result<Operation, RepositoryError> {
+    let linked_shipment = InvoiceRepository::new(connection).query_one(
+        InvoiceFilter::new().linked_invoice_id(EqualFilter::equal_to(&changelog_row.record_id)),
+    )?;
+
+    Ok(Operation::Delete { linked_shipment })
+}
+
+#[derive(Error, Debug)]
+#[error("Database error in processor ({0}) {1:?}")]
```

Comment:
side question: what is the `#[error` syntax? ( I'm having an issue where this might be helpful.. )

Reply:
I think it's from here... https://docs.rs/thiserror/latest/thiserror/
Does that help?

Reply:
It's easier to read display when the error is printed (and also the `Error` derive should capture the origin file/line). We don't quite have a consistent and descriptive spec of how/when to use it yet, still figuring it out